src/Microsoft.Health.Fhir.Core/Features/Search/Expressions/Parsers/SearchValueExpressionBuilderHelper.cs

```diff
@@ -271,6 +271,34 @@ void ISearchValueVisitor.Visit(TokenSearchValue token)
                         Expression.StringEquals(FieldName.TokenCode, _componentIndex, token.Code, false));
                 }
             }
+            else if (_modifier == SearchModifierCode.Not)
+            {
+                // Based on spec http://hl7.org/fhir/search.html#token,
+                // we need to make sure to test if system is missing or not based on how it is supplied.
+                if (token.System == null)
+                {
+                    // If the system is not supplied, then the token code is matched irrespective of the value of system.
+                    _outputExpression = Expression.NotContains(FieldName.TokenCode, _componentIndex, token.Code, false);
+                }
+                else if (token.System.Length == 0)
+                {
+                    // If the system is empty, then the token is matched if there is no system property.
+                    _outputExpression = Expression.And(
+                        Expression.Missing(FieldName.TokenSystem, _componentIndex),
+                        Expression.NotContains(FieldName.TokenCode, _componentIndex, token.Code, false));
```

Comment:
All other expression is do opposite of absence of modifier.
but not this one.
`not (a and b)` should be: `not(a) or not(b)`
but in this case it becomes: `(a) and not(b)`

shouldn't this be Expression.Or (Expression.Exist(), Expression.NotContains)?
I also surprised absence of `StringOperator.NotEquals`

Reply:
I wish there was a StringOperator.NotEquals. You're right, this should have been fully reversed, I'll fix that.

Reply:
Actually, disregard my comment, it's valid for one dimensional case, but here you have multidimensional case for system and code.
I probably shouldn't but out of curiosity I start implementing NotEquals, and I'll try to post PR with changes within day.

---
In reply to: [526329666](https://github.com/microsoft/fhir-server/pull/1462#discussion_r526329666) [](ancestors = 526329666)

Reply:
I think your comment is still accurate for this case. For the case of not "|<code>" it is asking for if it isn't equal to the code or it has a system. Thanks for looking into the NotEquals expression.

Reply:
Replace the entire section with the new Not expression